File: Misc/108_lc_2641.java
package Misc;

import java.util.*;
/*
 * 2641. Cousins in Binary Tree II
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public TreeNode replaceValueInTree(TreeNode root) {
        root.val = 0;
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        while(!q.isEmpty()){
            int nextLevelSum = 0;
            int qSize = q.size();
            List<TreeNode> parents = new ArrayList<>(q);
            for (int i=0; i<qSize; ++i) {
                TreeNode node = q.poll();
                if (node.left != null) {
                    q.offer(node.left);
                    nextLevelSum += node.left.val;
                }
                if (node.right != null) {
                    q.offer(node.right);
                    nextLevelSum += node.right.val;
                }
            }
            
            // iterating over all the values present in q as of now
            for (TreeNode n: parents) {
                int tempNextLevelSum = nextLevelSum;
                tempNextLevelSum -= n.left != null ? n.left.val : 0;
                tempNextLevelSum -= n.right != null ? n.right.val : 0;

                if (n.left != null) {n.left.val = tempNextLevelSum;}
                if (n.right != null) {n.right.val = tempNextLevelSum;}
            }
        }
        return root;
    }
}

File: Misc/108_lc_2641.py
"""
2641. Cousins in Binary Tree II
"""

from typing import List, Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        dummyRoot = root
        dq = deque()
        dq.append((dummyRoot, None))
        hmap = {}  # level : {parent: totalChildSum}

        level = 0
        while dq:
            for _ in range(len(dq)):
                node, parent = dq.popleft()
                if node.left:
                    dq.append((node.left, node))
                if node.right:
                    dq.append((node.right, node))

                leftNodeVal = node.left.val if node.left else 0
                rightNodeVal = node.right.val if node.right else 0
                valueToUpdate = rightNodeVal + leftNodeVal
                try:
                    hmap[level + 1].update({node: valueToUpdate})
                    hmap[level + 1]["total"] = (
                        hmap.get(level + 1).get("total") + valueToUpdate
                    )
                except:
                    hmap.update({level + 1: {node: valueToUpdate}})
                    hmap[level + 1].update({"total": valueToUpdate})

                if level == 0 or level == 1:
                    node.val = 0
                else:
                    node.val = hmap[level]["total"] - hmap[level][parent]
            level += 1
        return root
